### DISCUSSION TITLE
USB hub support

### DISCUSSION THREAD

**Discussion by Ionic** [MoleMan1024/audiowagon#140] [Q&A] [answered]
The documentation explicitly mentions that it's not possible to attach multiple media via a USB hub, but that doesn't answer questions about USB hubs in general.

As far as I understood, newer Polestar cars only have USB C slots, while older one feature either a mix of both USB A and C or only A.

Media with A plugs cannot be directly inserted into C sockets, of course, so you'll either need a passive adapter or a hub.

Has anyone ever tested attaching an A-plugged flash drive to a USB C hub and using that with AudioWagon?

On a related note, most USB C hubs also support SD cards, would that work with AudioWagon?

I know that Android itself supports this just fine (my mobile phone has no issues accessing either SD cards or flash drives connected to the hub), but Android Automotive OS as used by Volvo might behave differently...

**Comment by MoleMan1024** [comment, marked as answer]
@Ionic : there was never a Polestar car that had a USB-A slot in the front, they always had USB-C slots only. Most people use adapters to go from USB-C to USB-A. A few people use USB-C flash drives.

I think someone has tried a USB hub before and it is working if you plug in e.g. a mobile phone and a USB flash drive (just not two USB flash drives).

SD card reader support is mentioned [here](https://moleman1024.github.io/audiowagon/#limitations), you will have to try if it works or not.

If it is working in another version of Android, it is quite likely it will work in AAOS also.

**Comment by Ionic** [comment]
Oh, you're right... on two fronts.

While doing my homework, I stumbled across a manual page detailing [USB A sockets](https://web.archive.org/https://www.polestar.com/us/manual/polestar-1/2020/article/ec98b1ba8a08362ec0a8015149101eed/), but I failed to notice that this only holds for Polestar 1 models, which are very special plugin-hybrid cars that never made it to mass production. The Polestar 2, on the contrary, never had USB A sockets. My bad.

Okay, good! I only want to use the USB C hub for plugging in a USB A flash drive, so let's hope that it works. I'm not sure how low-level the library AudioWagon uses to read the data goes, but I'll just hope that basic USB functionality is provided through the system and that USB hubs will just work transparently.

As to SD cards... yeah, I've read that passage before, but failed to realize that memory card adapters are typically just USB hubs internally anyway, so also failed to deduce that what I had in mind worked for some people, but failed for other people.

I'll try a passive USB A to C adapter, as well as an USB C hub with both a flash drive and an SD card next Friday... if I actually get the car and if I can actually install AudioWagon on a rental car. :)

**Comment by micksmothers** [comment]
I have used a USB type A through a passive type C adapter successfully on my Polestar 2, however I ended up getting a USB drive that has both a USB A and C connector on it. Perfect for hooking to my PC to edit my song catalog and then a minimal footprint in the car. I also have a 90 degree connector for the USB port in the car so the drive lies flat. Hope this helps - if you're just getting a rental then it might now be worth making the investment though.

**Comment by Ionic** [comment]
I had the chance to test multiple things in a Polestar 2:

  - 1 TB flashdrive connected via
    - a passive USB A to USB C adapter
    - an USB C hub
  - SD(HC) card connected via
    - an old USB A 2.0 card reader connected via a passive USB A to USB C adapter
    - an USB C hub with an integrated card reader

All combinations worked flawlessly.

**Comment by mekp-code** [comment]
some new ideas devlopment

**Comment by mekp-code** [comment]
1 -  Allows Audiowagon to access ...USB...
    - Don't ask at every startup, but save it once and for all
    - Ability to change the above setting in the settings

2 - Display all folders and subfolders of each played MP3 file
    - Ability to change this display setting in the settings

3 - Ability to automatically advance to the next folder when playback of a folder is complete
    - Ability to change this automatic advance to the next folder setting in the settings

**Reply by MoleMan1024** [reply]
@mekp-code : The current discussion thread is about USB hubs only, please limit discussion to this topic.

Regarding your points:
1) is explained [in the FAQ page](https://moleman1024.github.io/audiowagon/faq.html#why-does-the-permission-popup-appear-each-time-the-always-open-audiowagon--checkbox-does-not-do-anything)
2) Sorry I don't understand what you mean. Please open a new discussion or feature request and explain in more detail
3) This has already been requested in issue #163 . Please comment on that issue instead

**Reply by mekp-code** [reply]
for the 2nd point :
foreach mp3 played, replace in the display the basename of the file played with the complete fielename

for example :
01Ouvert.mp3
02Couran.mp3
03Gavott.mp3
...
->
47-Bach\OuvFran\01Ouvert.mp3
47-Bach\OuvFran\02Couran.mp3
47-Bach\OuvFran\03Gavott.mp3
...

and ability to change this display setting in the parameters of Audiowagon

**Reply by MoleMan1024** [reply]
I have opened a new feature request for your 2nd point here https://github.com/MoleMan1024/audiowagon/issues/196 . Please discuss it further in there.
